tools/tcpretrans.py

```diff
@@ -70,6 +74,24 @@
 };
 BPF_PERF_OUTPUT(ipv6_events);
 
+// separate flow keys per address family
+struct ipv4_flow_key_t {
+    // XXX: switch some to u32's when supported
+    u32 saddr;
+    u32 daddr;
+    u16 lport;
+    u16 dport;
+};
+BPF_HASH(ipv4_count, struct ipv4_flow_key_t);
+
+struct ipv6_flow_key_t {
+    unsigned __int128 saddr;
+    unsigned __int128 daddr;
+    u64 lport;
+    u64 dport;
+};
```

Comment:
With the latest trunk, you can use "u16" for lport/dport. This is consistent with the type in the packet. Could you give a try and if it works, use "u16" instead of "u64"?

Reply:
Ergo, works kind of, though seems crashing in some free mechanism ... will leave at u64 until resolved.

# python tools/tcpretrans.py -c
Tracing retransmits ... Hit Ctrl-C to end
^C
LADDR:LPORT              RADDR:RPORT          RETRANSMITS
2003:62:4643:bd23:d1eb:a5d9:3fab:554a:54190 <-> 2003:62:4643:bd23:d1eb:a5d9:3fab:54a1:190 1         
2003:62:4643:bd23:d1eb:a5d9:3fab:554a:54188 <-> 2003:62:4643:bd23:d1eb:a5d9:3fab:54a1:190 1         
*** Error in `python': free(): invalid next size (fast): 0x0000000000fc7290 ***
======= Backtrace: =========
/lib/x86_64-linux-gnu/libc.so.6(+0x777e5)[0x7fcf6d4de7e5]
/lib/x86_64-linux-gnu/libc.so.6(+0x8037a)[0x7fcf6d4e737a]
/lib/x86_64-linux-gnu/libc.so.6(cfree+0x4c)[0x7fcf6d4eb53c]
/usr/lib/python2.7/lib-dynload/_ctypes.x86_64-linux-gnu.so(+0x13e14)[0x7fcf6c4d8e14]
python[0x4fd90a]
python[0x4b5244]

Reply:
Did you rebase against latest trunk?

Reply:
Yes, feel free to verify by checking out my PR.

E.g. that's in ...

$ git blame tests/python/test_clang.py | grep -i Yong
538a84e1 tests/python/test_clang.py (Yonghong Song   2017-12-05 00:18:48 -0800 678)     def test_padding_types(self):
538a84e1 tests/python/test_clang.py (Yonghong Song   2017-12-05 00:18:48 -0800 679)         text = """
538a84e1 tests/python/test_clang.py (Yonghong Song   2017-12-05 00:18:48 -0800 680) struct key_t {
538a84e1 tests/python/test_clang.py (Yonghong Song   2017-12-05 00:18:48 -0800 681)   u32 f1_1;               /* offset 0 */

Reply:
Thanks. Let me try then.

Reply:
@cherusk Actually your change is not the top of the trunk.
```
commit 538a84e1f821a97468a55d0e97a5c0a4617a1271
Author: Yonghong Song <yhs@fb.com>
Date:   Tue Dec 5 00:18:48 2017 -0800

    provide padded structure for table_{key|leaf}_desc API
```
corresponds to the above commit.
You should have the following one
```
commit b32b4a5fffb61ae6fd5e75e30c06763ade8ed1a4
Author: Yonghong Song <yhs@fb.com>
Date:   Mon Dec 11 20:26:38 2017 -0800

    add padding to the end of structure
```
to make it work.

Reply:
Believe me ... I intreat you ...

$ git log -n 10 | grep commit -A 4
commit 5a1aa35c618d67cab0a0867daaad7a68c121c16a
Author: Matthias Tafelmeier <matthias.tafelmeier@gmx.net>
Date:   Wed Dec 13 18:50:22 2017 +0100

*-->>>>    tcpretrans: extend for counting*
--
commit e5db52bfd61497dfcd310f88f6dc14c36b9b2aba
Merge: b384b76 b32b4a5
Author: 4ast <alexei.starovoitov@gmail.com>
Date:   Tue Dec 12 15:08:59 2017 -0800

--
commit b384b76e2cefbac2793473e4c97843318fe75763
Merge: 2faf060 3c399da
Author: Brendan Gregg <brendan.d.gregg@gmail.com>
Date:   Tue Dec 12 11:35:21 2017 -0800

--
commit 2faf0605489447245d175e51350feecfac862383
Merge: 34b3c10 538a84e
Author: 4ast <alexei.starovoitov@gmail.com>
Date:   Tue Dec 12 09:57:18 2017 -0800

--
commit b32b4a5fffb61ae6fd5e75e30c06763ade8ed1a4
Author: Yonghong Song <yhs@fb.com>
Date:   Mon Dec 11 20:26:38 2017 -0800

    add padding to the end of structure
--
commit 3c399da1caa238f2f16a6f13f8b98f51d4d74d79
Author: htbegin <hotforest@gmail.com>
Date:   Sat Dec 9 09:34:56 2017 +0800

    execsnoop: fix the prototype for kprobe__sys_execve
--
commit 34b3c10e69ee7822a047f60850e124dbe2b5aa43
Merge: bf1a54c c669257
Author: yonghong-song <ys114321@gmail.com>
Date:   Thu Dec 7 15:34:58 2017 -0800

--
commit c669257c1f16eae4c695866bea147f57a9092b80
Author: Teng Qin <qinteng@fb.com>
Date:   Tue Dec 5 17:06:57 2017 -0800

    Do not print to stderr if user specified log buffer
--
commit 538a84e1f821a97468a55d0e97a5c0a4617a1271
Author: Yonghong Song <yhs@fb.com>
Date:   Tue Dec 5 00:18:48 2017 -0800

*-->>>>     provide padded structure for table_{key|leaf}_desc API*

Reply:
Somehow I cannot reproduce the failure.

```
-bash-4.2$ sudo ./tcpretrans.py
Tracing retransmits ... Hit Ctrl-C to end
TIME     PID    IP LADDR:LPORT          T> RADDR:RPORT          STATE
08:47:57 0      6  2401:db00:11:d0af:face:0:2d:0:52436 R> 2401:db00:12:3f23:face:b00c:0:35f4:88 ESTABLISHED
08:47:57 0      6  2401:db00:11:d0af:face:0:2d:0:52436 R> 2401:db00:12:3f23:face:b00c:0:35f4:88 ESTABLISHED
08:48:11 0      6  2401:db00:11:d0af:face:0:2d:0:40796 R> 2401:db00:30:3f0e:face:b00c:0:35f4:88 ESTABLISHED
08:48:11 0      6  2401:db00:11:d0af:face:0:2d:0:40796 R> 2401:db00:30:3f0e:face:b00c:0:35f4:88 ESTABLISHED
^C-bash-4.2$ git diff
diff --git a/tools/tcpretrans.py b/tools/tcpretrans.py
old mode 100644
new mode 100755
index b599d55..a685fde
--- a/tools/tcpretrans.py
+++ b/tools/tcpretrans.py
@@ -86,8 +86,8 @@ BPF_HASH(ipv4_count, struct ipv4_flow_key_t);
 struct ipv6_flow_key_t {
     unsigned __int128 saddr;
     unsigned __int128 daddr;
-    u64 lport;
-    u64 dport;
+    u16 lport;
+    u16 dport;
 };
 BPF_HASH(ipv6_count, struct ipv6_flow_key_t);
 
-bash-4.2$ 
```

Reply:
Some quick tests here from latest trunk seem to work fine

>  sudo python ./retrans.py -c
Tracing retransmits ... Hit Ctrl-C to end
^C
LADDR:LPORT              RADDR:RPORT          RETRANSMITS
2001:6a8:308f:2:f956:c16a:73c:360e:50288 <-> 202:ddd::ae:1235     2
2001:6a8:308f:2:f956:c16a:73c:360e:41590 <-> 200:::1234           2
2001:6a8:308f:2:f956:c16a:73c:360e:60820 <-> 200:::1235           2
2001:6a8:308f:2:f956:c16a:73c:360e:50932 <-> 202::ae:1235         2

Did you rebuild/install bcc (i.e. make && make install) ?

As an extra detail, could you put the IPv6 addresses between brackets so that we can more easily distinguish the port number (this matches common practice from other software), e.g. display this instead:
> Tracing retransmits ... Hit Ctrl-C to end
^C
LADDR:LPORT              RADDR:RPORT          RETRANSMITS
[2001:6a8:308f:2::]:50288 <-> [202:ddd::ae]:1235     2
[2001:6a8:308f:2::]:41590 <-> [200::]:1234           2
[2001:6a8:308f:2::]:60820 <-> [200::]:1235           2
[2001:6a8:308f:2::]:50932 <-> [202::ae]:1235         2

Reply:
A much lengthy try with scp 1.5G data and no issues.
```
-bash-4.2$ sudo ./tcpretrans.py
Tracing retransmits ... Hit Ctrl-C to end
TIME     PID    IP LADDR:LPORT          T> RADDR:RPORT          STATE
08:53:15 0      6  2401:db00:11:d0af:face:0:2d:0:22 R> 2620:10d:c082:1055:ac25:6915:257e:f98e:54515 ESTABLISHED
08:53:16 2357527 6  2401:db00:11:d0af:face:0:2d:0:22 R> 2620:10d:c082:1055:ac25:6915:257e:f98e:54515 ESTABLISHED
08:53:16 2357527 6  2401:db00:11:d0af:face:0:2d:0:22 R> 2620:10d:c082:1055:ac25:6915:257e:f98e:54515 ESTABLISHED
08:53:16 2357527 6  2401:db00:11:d0af:face:0:2d:0:22 R> 2620:10d:c082:1055:ac25:6915:257e:f98e:54515 ESTABLISHED
08:53:16 2357527 6  2401:db00:11:d0af:face:0:2d:0:22 R> 2620:10d:c082:1055:ac25:6915:257e:f98e:54515 ESTABLISHED
08:53:16 2357527 6  2401:db00:11:d0af:face:0:2d:0:22 R> 2620:10d:c082:1055:ac25:6915:257e:f98e:54515 ESTABLISHED
08:53:16 2357527 6  2401:db00:11:d0af:face:0:2d:0:22 R> 2620:10d:c082:1055:ac25:6915:257e:f98e:54515 ESTABLISHED
08:53:16 2357527 6  2401:db00:11:d0af:face:0:2d:0:22 R> 2620:10d:c082:1055:ac25:6915:257e:f98e:54515 ESTABLISHED
08:53:16 2357527 6  2401:db00:11:d0af:face:0:2d:0:22 R> 2620:10d:c082:1055:ac25:6915:257e:f98e:54515 ESTABLISHED
08:53:16 2357527 6  2401:db00:11:d0af:face:0:2d:0:22 R> 2620:10d:c082:1055:ac25:6915:257e:f98e:54515 ESTABLISHED
08:53:16 2357527 6  2401:db00:11:d0af:face:0:2d:0:22 R> 2620:10d:c082:1055:ac25:6915:257e:f98e:54515 ESTABLISHED
08:53:16 2357527 6  2401:db00:11:d0af:face:0:2d:0:22 R> 2620:10d:c082:1055:ac25:6915:257e:f98e:54515 ESTABLISHED
08:53:16 2357527 6  2401:db00:11:d0af:face:0:2d:0:22 R> 2620:10d:c082:1055:ac25:6915:257e:f98e:54515 ESTABLISHED
08:53:16 2357527 6  2401:db00:11:d0af:face:0:2d:0:22 R> 2620:10d:c082:1055:ac25:6915:257e:f98e:54515 ESTABLISHED
......
08:53:30 0      6  2401:db00:11:d0af:face:0:2d:0:22 R> 2620:10d:c082:1055:ac25:6915:257e:f98e:54515 ESTABLISHED
08:53:30 0      6  2401:db00:11:d0af:face:0:2d:0:22 R> 2620:10d:c082:1055:ac25:6915:257e:f98e:54515 ESTABLISHED
08:53:30 0      6  2401:db00:11:d0af:face:0:2d:0:22 R> 2620:10d:c082:1055:ac25:6915:257e:f98e:54515 ESTABLISHED
08:53:30 0      6  2401:db00:11:d0af:face:0:2d:0:22 R> 2620:10d:c082:1055:ac25:6915:257e:f98e:54515 ESTABLISHED
08:53:30 0      6  2401:db00:11:d0af:face:0:2d:0:22 R> 2620:10d:c082:1055:ac25:6915:257e:f98e:54515 ESTABLISHED
08:53:30 0      6  2401:db00:11:d0af:face:0:2d:0:22 R> 2620:10d:c082:1055:ac25:6915:257e:f98e:54515 ESTABLISHED
08:53:30 0      6  2401:db00:11:d0af:face:0:2d:0:22 R> 2620:10d:c082:1055:ac25:6915:257e:f98e:54515 ESTABLISHED
08:53:31 0      6  2401:db00:11:d0af:face:0:2d:0:22 R> 2620:10d:c082:1055:ac25:6915:257e:f98e:54515 ESTABLISHED
08:53:31 3416497 6  2401:db00:11:d0af:face:0:2d:0:22 R> 2620:10d:c082:1055:ac25:6915:257e:f98e:54515 ESTABLISHED
08:53:31 3416497 6  2401:db00:11:d0af:face:0:2d:0:22 R> 2620:10d:c082:1055:ac25:6915:257e:f98e:54515 ESTABLISHED
08:53:31 3416497 6  2401:db00:11:d0af:face:0:2d:0:22 R> 2620:10d:c082:1055:ac25:6915:257e:f98e:54515 ESTABLISHED
^C-bash-4.2$
```

Reply:
This is with the count:
```
-bash-4.2$ sudo ./tcpretrans.py -c
Tracing retransmits ... Hit Ctrl-C to end
^C
LADDR:LPORT              RADDR:RPORT          RETRANSMITS
2401:db00:11:d0af:face:0:2d:0:22 <-> 2620:10d:c082:1055:ac25:6915:257e:f98e:54570 450       
-bash-4.2$ git diff tcpretrans.py
diff --git a/tools/tcpretrans.py b/tools/tcpretrans.py
old mode 100644
new mode 100755
index b599d55..a685fde
--- a/tools/tcpretrans.py
+++ b/tools/tcpretrans.py
@@ -86,8 +86,8 @@ BPF_HASH(ipv4_count, struct ipv4_flow_key_t);
 struct ipv6_flow_key_t {
     unsigned __int128 saddr;
     unsigned __int128 daddr;
-    u64 lport;
-    u64 dport;
+    u16 lport;
+    u16 dport;
 };
 BPF_HASH(ipv6_count, struct ipv6_flow_key_t);
 
-bash-4.2$ 
```

Reply:
@oliviertilmans Valid point as to the output: did so e.g. here  https://github.com/cherusk/fleutan - only to illustrate that I'm not ignorant to that topic/have suffered enough from ss (iproute) - please, bring forward to netdev as well ...

@yonghong-song Thanks for validating.